blockchain/utxocache.go

```diff
@@ -0,0 +1,987 @@
+	utxoView
+
+	// getEntries attempts to fetch a serise of entries from the UTXO view.
+	// If a single entry is not found within the view, then an error is
+	// returned.
+	getEntries(outpoints map[wire.OutPoint]struct{}) (map[wire.OutPoint]*UtxoEntry, error)
+}
+
+// utxoCache is a cached utxo view in the chainstate of a BlockChain.
+//
+// It implements the utxoView interface, but should only be used as such with the
+// state mutex held.  It also implements the utxoByHashSource interface.
+type utxoCache struct {
+	db database.DB
+
+	// maxTotalMemoryUsage is the maximum memory usage in bytes that the state
+	// should contain in normal circumstances.
+	maxTotalMemoryUsage uint64
+
+	// This mutex protects the internal state.
+	// A simple mutex instead of a read-write mutex is chosen because the main
+	// read method also possibly does a write on a cache miss.
+	mtx sync.Mutex
+
+	// cachedEntries keeps the internal cache of the utxo state.  The tfModified
+	// flag indicates that the state of the entry (potentially) deviates from the
+	// state in the database.  Explicit nil values in the map are used to
+	// indicate that the database does not contain the entry.
+	cachedEntries    map[wire.OutPoint]*UtxoEntry
+	totalEntryMemory uint64 // Total memory usage in bytes.
+	lastFlushHash    chainhash.Hash
+}
+
+// newUtxoCache initiates a new utxo cache instance with its memory usage limited
+// to the given maximum.
+func newUtxoCache(db database.DB, maxTotalMemoryUsage uint64) *utxoCache {
+	avgEntrySize := outpointSize + 8 + baseEntrySize + pubKeyHashLen
+	numMaxElements := maxTotalMemoryUsage / avgEntrySize
+	log.Info("Pre-alloacting for %v entries: ", numMaxElements)
+
+	return &utxoCache{
+		db:                  db,
+		maxTotalMemoryUsage: maxTotalMemoryUsage,
+		cachedEntries:       make(map[wire.OutPoint]*UtxoEntry, numMaxElements),
+	}
+}
+
+// totalMemoryUsage returns the total memory usage in bytes of the UTXO cache.
+//
+// This method should be called with the state lock held.
+func (s *utxoCache) totalMemoryUsage() uint64 {
+	// Total memory is all the keys plus the total memory of all the entries.
+	nbEntries := uint64(len(s.cachedEntries))
+
+	// Total size is total size of the keys + total size of the pointers in the
+	// map + total size of the elements held in the pointers.
+	return nbEntries*outpointSize + nbEntries*8 + s.totalEntryMemory
+}
+
+// TotalMemoryUsage returns the total memory usage in bytes of the UTXO cache.
+//
+// This method is safe for concurrent access.
+func (s *utxoCache) TotalMemoryUsage() uint64 {
+	s.mtx.Lock()
+	tmu := s.totalMemoryUsage()
+	s.mtx.Unlock()
+	return tmu
+}
+
+// seekAndCacheEntries attempts to fetch a series of entries from the
+// database.  In any aren't found, then nil is returned. All entries found are
+// cached in the process.
+//
+// This method should be called with the state lock held.
+func (s *utxoCache) seekAndCacheEntries(ops ...wire.OutPoint) (
```

Comment:
Can we reduce code duplication between `seekAndCacheEntries` and `fetchAndCacheEntries`?  I'm not sure where `fetchAndCacheEntries` is used.